The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7906925198/job/21582833759) [(plain)](https://github.com/rust-lang/rust/commit/00345ad3447734bf737358e10fa2ae6361313359/checks/21582833759/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_f99e66e6-6a8a-4ac9-94d8-6d2d54dc34b2
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=TyKind-Err-guar-2
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_f99e66e6-6a8a-4ac9-94d8-6d2d54dc34b2
GITHUB_REF=refs/pull/121109/merge
GITHUB_REF_NAME=121109/merge
GITHUB_REF_PROTECTED=false
---
2 -     if bar && if !baz {
2 +     if bar && !baz {
  |

note: no errors encountered even though delayed bugs were created

note: those delayed bugs will now be shown as internal compiler errors

error: internal compiler error: obsolete `..` syntax
    |
    |
167 | impl T for  .. {
    |
note: delayed at compiler/rustc_parse/src/parser/item.rs:594:35 - disabled backtrace
   --> tests/source/type.rs:167:13
    |
    |
167 | impl T for  .. {


note: no errors encountered even though delayed bugs were created

note: those delayed bugs will now be shown as internal compiler errors

error: internal compiler error: obsolete `..` syntax
##[error]   --> tests/target/type.rs:175:12
175 | impl T for .. {}
    |            ^^
    |
note: delayed at compiler/rustc_parse/src/parser/item.rs:594:35 - disabled backtrace
---

failures:

---- test::system_tests stdout ----
2024-02-14T20:38:59.848866Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:38:59.853509Z  WARN rustfmt_nightly::test: Default value 2015 used explicitly for edition
2024-02-14T20:38:59.865757Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:38:59.893972Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:38:59.949042Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.038084Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.038089Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.041504Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
Warning: the `merge_imports` option is deprecated. Use `imports_granularity="Crate"` instead
2024-02-14T20:39:00.290793Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.306817Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.318343Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.452170Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.454341Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_imports
2024-02-14T20:39:00.558141Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_imports
2024-02-14T20:39:00.573401Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.579933Z  WARN rustfmt_nightly::test: Default value true used explicitly for empty_item_single_line
2024-02-14T20:39:00.586439Z  WARN rustfmt_nightly::test: Default value false used explicitly for hard_tabs
2024-02-14T20:39:00.591416Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.591421Z  WARN rustfmt_nightly::test: Default value false used explicitly for format_strings
2024-02-14T20:39:00.597732Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_modules
2024-02-14T20:39:00.608079Z  WARN rustfmt_nightly::test: Default value 4 used explicitly for tab_spaces
2024-02-14T20:39:00.608084Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.610356Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.614752Z  WARN rustfmt_nightly::test: Default value false used explicitly for fn_single_line
2024-02-14T20:39:00.621764Z  WARN rustfmt_nightly::test: Default value Vertical used explicitly for trailing_comma
2024-02-14T20:39:00.623959Z  WARN rustfmt_nightly::test: Default value true used explicitly for format_macro_bodies
2024-02-14T20:39:00.630941Z  WARN rustfmt_nightly::test: Default value false used explicitly for normalize_doc_attributes
2024-02-14T20:39:00.635722Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.635727Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:00.640097Z  WARN rustfmt_nightly::test: Default value false used explicitly for reorder_impl_items
2024-02-14T20:39:00.642363Z  WARN rustfmt_nightly::test: Default value 100 used explicitly for doc_comment_code_block_width
2024-02-14T20:39:00.644860Z  WARN rustfmt_nightly::test: Default value 100 used explicitly for doc_comment_code_block_width
2024-02-14T20:39:00.654450Z  WARN rustfmt_nightly::test: Default value false used explicitly for force_multiline_blocks
2024-02-14T20:39:00.659057Z  WARN rustfmt_nightly::test: Default value false used explicitly for condense_wildcard_suffixes
2024-02-14T20:39:00.663728Z  WARN rustfmt_nightly::test: Default value 80 used explicitly for comment_width
2024-02-14T20:39:00.665905Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:00.668029Z  WARN rustfmt_nightly::test: Default value true used explicitly for format_generated_files
2024-02-14T20:39:00.672275Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_imports
2024-02-14T20:39:00.678587Z  WARN rustfmt_nightly::test: Default value false used explicitly for disable_all_formatting
2024-02-14T20:39:00.689992Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.692475Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.697344Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.699735Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.704179Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.706473Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.706478Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.706480Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.708726Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.708732Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.710967Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.713216Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.719802Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.722001Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.728554Z  WARN rustfmt_nightly::test: Default value false used explicitly for normalize_comments
2024-02-14T20:39:00.738102Z  WARN rustfmt_nightly::test: Default value SameLineWhere used explicitly for brace_style
2024-02-14T20:39:00.744710Z  WARN rustfmt_nightly::test: Default value SameLineWhere used explicitly for brace_style
2024-02-14T20:39:00.753821Z  WARN rustfmt_nightly::test: Default value 10 used explicitly for short_array_element_width_threshold
2024-02-14T20:39:00.762080Z  WARN rustfmt_nightly::test: Default value 50 used explicitly for single_line_let_else_max_width
2024-02-14T20:39:00.765126Z  WARN rustfmt_nightly::test: Default value true used explicitly for match_arm_blocks
2024-02-14T20:39:00.769729Z  WARN rustfmt_nightly::test: Default value Tall used explicitly for fn_params_layout
2024-02-14T20:39:00.781448Z  WARN rustfmt_nightly::test: Default value Default used explicitly for use_small_heuristics
2024-02-14T20:39:00.783847Z  WARN rustfmt_nightly::test: Default value true used explicitly for merge_derives
2024-02-14T20:39:00.786224Z  WARN rustfmt_nightly::test: Default value true used explicitly for force_explicit_abi
2024-02-14T20:39:00.792882Z  WARN rustfmt_nightly::test: Default value false used explicitly for spaces_around_ranges
2024-02-14T20:39:00.795312Z  WARN rustfmt_nightly::test: Default value Mixed used explicitly for imports_layout
2024-02-14T20:39:00.795317Z  WARN rustfmt_nightly::test: Default value Block used explicitly for imports_indent
2024-02-14T20:39:00.799705Z  WARN rustfmt_nightly::test: Default value false used explicitly for use_try_shorthand
2024-02-14T20:39:00.804300Z  WARN rustfmt_nightly::test: Default value Never used explicitly for match_arm_leading_pipes
2024-02-14T20:39:00.831434Z  WARN rustfmt_nightly::test: Default value false used explicitly for use_field_init_shorthand
2024-02-14T20:39:00.833722Z  WARN rustfmt_nightly::test: Default value Wide used explicitly for type_punctuation_density
2024-02-14T20:39:00.838417Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.843180Z  WARN rustfmt_nightly::test: Default value false used explicitly for format_macro_matchers
2024-02-14T20:39:00.845782Z  WARN rustfmt_nightly::test: Default value true used explicitly for remove_nested_parens
2024-02-14T20:39:00.850160Z  WARN rustfmt_nightly::test: Default value false used explicitly for match_block_trailing_comma
2024-02-14T20:39:00.856833Z  WARN rustfmt_nightly::test: Default value AlwaysSameLine used explicitly for control_brace_style
2024-02-14T20:39:00.865991Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.875119Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.884092Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
Warning: the `fn_args_layout` option is deprecated. Use `fn_params_layout`. instead
2024-02-14T20:39:01.100784Z  WARN rustfmt_nightly::test: Default value Block used explicitly for imports_indent
2024-02-14T20:39:01.100789Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:01.119036Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_imports
2024-02-14T20:39:01.154996Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_modules
2024-02-14T20:39:01.199958Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:01.239157Z  WARN rustfmt_nightly::test: Default value [] used explicitly for skip_macro_invocations
2024-02-14T20:39:01.278958Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
Box<dyn Any>
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'test::system_tests' panicked at src/tools/rustfmt/src/test/mod.rs:76:10:
Failed to join a test thread: Any { .. }
Failed to join a test thread: Any { .. }

---- test::idempotence_tests stdout ----
2024-02-14T20:38:59.798530Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:38:59.867705Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:38:59.872350Z  WARN rustfmt_nightly::test: Default value 2015 used explicitly for edition
2024-02-14T20:38:59.885512Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:38:59.919558Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:38:59.996512Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.145635Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.145641Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.149079Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.158019Z  WARN rustfmt_nightly::test: Default value false used explicitly for format_macro_matchers
2024-02-14T20:39:00.177979Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
Warning: the `merge_imports` option is deprecated. Use `imports_granularity="Crate"` instead
2024-02-14T20:39:00.487318Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.493719Z  WARN rustfmt_nightly::test: Default value 0 used explicitly for struct_field_align_threshold
2024-02-14T20:39:00.496097Z  WARN rustfmt_nightly::test: Default value false used explicitly for hard_tabs
2024-02-14T20:39:00.496103Z  WARN rustfmt_nightly::test: Default value 4 used explicitly for tab_spaces
2024-02-14T20:39:00.496113Z  WARN rustfmt_nightly::test: Default value SameLineWhere used explicitly for brace_style
2024-02-14T20:39:00.496116Z  WARN rustfmt_nightly::test: Default value 100 used explicitly for max_width
2024-02-14T20:39:00.496118Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_modules
2024-02-14T20:39:00.510659Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.524141Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.631965Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:00.704526Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.706719Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_imports
2024-02-14T20:39:00.837545Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_imports
2024-02-14T20:39:00.849901Z  WARN rustfmt_nightly::test: Default value Block used explicitly for imports_indent
2024-02-14T20:39:00.854344Z  WARN rustfmt_nightly::test: Default value SameLineWhere used explicitly for brace_style
2024-02-14T20:39:00.856900Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.861540Z  WARN rustfmt_nightly::test: Default value true used explicitly for empty_item_single_line
2024-02-14T20:39:00.867902Z  WARN rustfmt_nightly::test: Default value false used explicitly for hard_tabs
2024-02-14T20:39:00.872825Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.872830Z  WARN rustfmt_nightly::test: Default value false used explicitly for format_strings
2024-02-14T20:39:00.879021Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_modules
2024-02-14T20:39:00.889282Z  WARN rustfmt_nightly::test: Default value true used explicitly for combine_control_expr
2024-02-14T20:39:00.889287Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.892128Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.894947Z  WARN rustfmt_nightly::test: Default value 4 used explicitly for tab_spaces
2024-02-14T20:39:00.894952Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.897203Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.901523Z  WARN rustfmt_nightly::test: Default value false used explicitly for fn_single_line
2024-02-14T20:39:00.908274Z  WARN rustfmt_nightly::test: Default value Vertical used explicitly for trailing_comma
2024-02-14T20:39:00.910446Z  WARN rustfmt_nightly::test: Default value true used explicitly for format_macro_bodies
2024-02-14T20:39:00.917263Z  WARN rustfmt_nightly::test: Default value false used explicitly for normalize_doc_attributes
2024-02-14T20:39:00.921630Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.921635Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:00.925939Z  WARN rustfmt_nightly::test: Default value false used explicitly for reorder_impl_items
2024-02-14T20:39:00.928124Z  WARN rustfmt_nightly::test: Default value 100 used explicitly for doc_comment_code_block_width
2024-02-14T20:39:00.930672Z  WARN rustfmt_nightly::test: Default value 100 used explicitly for doc_comment_code_block_width
2024-02-14T20:39:00.940114Z  WARN rustfmt_nightly::test: Default value false used explicitly for force_multiline_blocks
2024-02-14T20:39:00.944611Z  WARN rustfmt_nightly::test: Default value false used explicitly for condense_wildcard_suffixes
2024-02-14T20:39:00.948917Z  WARN rustfmt_nightly::test: Default value 80 used explicitly for comment_width
2024-02-14T20:39:00.951026Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:00.953144Z  WARN rustfmt_nightly::test: Default value true used explicitly for format_generated_files
2024-02-14T20:39:00.957325Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_imports
2024-02-14T20:39:00.963475Z  WARN rustfmt_nightly::test: Default value false used explicitly for disable_all_formatting
2024-02-14T20:39:00.974698Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.976954Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.980965Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.983355Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.987754Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.990039Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:00.990045Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.990048Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.992280Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.992285Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:00.994535Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.997460Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:00.999676Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:01.006270Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:01.008468Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:01.010693Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:01.017219Z  WARN rustfmt_nightly::test: Default value false used explicitly for normalize_comments
2024-02-14T20:39:01.026052Z  WARN rustfmt_nightly::test: Default value SameLineWhere used explicitly for brace_style
2024-02-14T20:39:01.032643Z  WARN rustfmt_nightly::test: Default value SameLineWhere used explicitly for brace_style
2024-02-14T20:39:01.041555Z  WARN rustfmt_nightly::test: Default value 10 used explicitly for short_array_element_width_threshold
2024-02-14T20:39:01.049543Z  WARN rustfmt_nightly::test: Default value 50 used explicitly for single_line_let_else_max_width
2024-02-14T20:39:01.052358Z  WARN rustfmt_nightly::test: Default value true used explicitly for match_arm_blocks
2024-02-14T20:39:01.056874Z  WARN rustfmt_nightly::test: Default value Tall used explicitly for fn_params_layout
2024-02-14T20:39:01.068563Z  WARN rustfmt_nightly::test: Default value Default used explicitly for use_small_heuristics
2024-02-14T20:39:01.070941Z  WARN rustfmt_nightly::test: Default value true used explicitly for merge_derives
2024-02-14T20:39:01.073280Z  WARN rustfmt_nightly::test: Default value true used explicitly for force_explicit_abi
2024-02-14T20:39:01.079915Z  WARN rustfmt_nightly::test: Default value false used explicitly for spaces_around_ranges
2024-02-14T20:39:01.082296Z  WARN rustfmt_nightly::test: Default value true used explicitly for trailing_semicolon
2024-02-14T20:39:01.086902Z  WARN rustfmt_nightly::test: Default value Block used explicitly for imports_indent
2024-02-14T20:39:01.089130Z  WARN rustfmt_nightly::test: Default value Block used explicitly for imports_indent
2024-02-14T20:39:01.089135Z  WARN rustfmt_nightly::test: Default value Mixed used explicitly for imports_layout
2024-02-14T20:39:01.091329Z  WARN rustfmt_nightly::test: Default value Mixed used explicitly for imports_layout
2024-02-14T20:39:01.091334Z  WARN rustfmt_nightly::test: Default value Block used explicitly for imports_indent
2024-02-14T20:39:01.095680Z  WARN rustfmt_nightly::test: Default value false used explicitly for use_try_shorthand
2024-02-14T20:39:01.100231Z  WARN rustfmt_nightly::test: Default value Never used explicitly for match_arm_leading_pipes
2024-02-14T20:39:01.124866Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_unformatted
2024-02-14T20:39:01.129314Z  WARN rustfmt_nightly::test: Default value false used explicitly for use_field_init_shorthand
2024-02-14T20:39:01.131560Z  WARN rustfmt_nightly::test: Default value Wide used explicitly for type_punctuation_density
2024-02-14T20:39:01.136261Z  WARN rustfmt_nightly::test: Default value false used explicitly for error_on_line_overflow
2024-02-14T20:39:01.141028Z  WARN rustfmt_nightly::test: Default value false used explicitly for format_macro_matchers
2024-02-14T20:39:01.143609Z  WARN rustfmt_nightly::test: Default value true used explicitly for remove_nested_parens
2024-02-14T20:39:01.147956Z  WARN rustfmt_nightly::test: Default value false used explicitly for match_block_trailing_comma
2024-02-14T20:39:01.154660Z  WARN rustfmt_nightly::test: Default value AlwaysSameLine used explicitly for control_brace_style
2024-02-14T20:39:01.163779Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:01.175115Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:01.183955Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
Warning: the `fn_args_layout` option is deprecated. Use `fn_params_layout`. instead
2024-02-14T20:39:01.246035Z  WARN rustfmt_nightly::test: Default value Preserve used explicitly for hex_literal_case
2024-02-14T20:39:01.248185Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:01.250414Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:01.252690Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:01.259176Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:01.274794Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:01.290452Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:01.292575Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:01.493080Z  WARN rustfmt_nightly::test: Default value Block used explicitly for indent_style
2024-02-14T20:39:01.493086Z  WARN rustfmt_nightly::test: Default value Block used explicitly for imports_indent
2024-02-14T20:39:01.510674Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_imports
2024-02-14T20:39:01.553993Z  WARN rustfmt_nightly::test: Default value true used explicitly for reorder_modules
2024-02-14T20:39:01.619055Z  WARN rustfmt_nightly::test: Default value false used explicitly for wrap_comments
2024-02-14T20:39:01.664244Z  WARN rustfmt_nightly::test: Default value [] used explicitly for skip_macro_invocations
2024-02-14T20:39:01.711093Z  WARN rustfmt_nightly::test: Default value One used explicitly for version
2024-02-14T20:39:01.751214Z  WARN rustfmt_nightly::test: Default value true used explicitly for format_macro_bodies
Box<dyn Any>
thread 'test::idempotence_tests' panicked at src/tools/rustfmt/src/test/mod.rs:76:10:
Failed to join a test thread: Any { .. }

```

</details>

--- a/compiler/rustc_ast/src/ast.rs
+++ b/compiler/rustc_ast/src/ast.rs
@@ -2136,10 +2136,12 @@ pub enum TyKind {
     ImplicitSelf,
     /// A macro in the type position.
     MacCall(P<MacCall>),
-    /// Placeholder for a kind that has failed to be defined.
-    Err,
     /// Placeholder for a `va_list`.
     CVarArgs,
+    /// Sometimes we need a dummy value when no error has occurred.
+    Dummy,
+    /// Placeholder for a kind that has failed to be defined.
+    Err(ErrorGuaranteed),
 }
 
 impl TyKind {

--- a/compiler/rustc_ast/src/mut_visit.rs
+++ b/compiler/rustc_ast/src/mut_visit.rs
@@ -481,7 +481,12 @@ pub fn noop_visit_ty<T: MutVisitor>(ty: &mut P<Ty>, vis: &mut T) {
     let Ty { id, kind, span, tokens } = ty.deref_mut();
     vis.visit_id(id);
     match kind {
-        TyKind::Infer | TyKind::ImplicitSelf | TyKind::Err | TyKind::Never | TyKind::CVarArgs => {}
+        TyKind::Infer
+        | TyKind::ImplicitSelf
+        | TyKind::Err(_)
+        | TyKind::Dummy
+        | TyKind::Never
+        | TyKind::CVarArgs => {}
         TyKind::Slice(ty) => vis.visit_ty(ty),
         TyKind::Ptr(mt) => vis.visit_mt(mt),
         TyKind::Ref(lt, mt) => {
@@ -1649,7 +1654,7 @@ impl DummyAstNode for Ty {
     fn dummy() -> Self {
         Ty {
             id: DUMMY_NODE_ID,
-            kind: TyKind::Err,
+            kind: TyKind::Dummy,
             span: Default::default(),
             tokens: Default::default(),
         }

--- a/compiler/rustc_ast/src/visit.rs
+++ b/compiler/rustc_ast/src/visit.rs
@@ -447,7 +447,7 @@ pub fn walk_ty<'a, V: Visitor<'a>>(visitor: &mut V, typ: &'a Ty) {
             walk_list!(visitor, visit_param_bound, bounds, BoundKind::Impl);
         }
         TyKind::Typeof(expression) => visitor.visit_anon_const(expression),
-        TyKind::Infer | TyKind::ImplicitSelf | TyKind::Err => {}
+        TyKind::Infer | TyKind::ImplicitSelf | TyKind::Dummy | TyKind::Err(_) => {}
         TyKind::MacCall(mac) => visitor.visit_mac_call(mac),
         TyKind::Never | TyKind::CVarArgs => {}
         TyKind::AnonStruct(_, ref fields) | TyKind::AnonUnion(_, ref fields) => {

--- a/compiler/rustc_ast_lowering/src/lib.rs
+++ b/compiler/rustc_ast_lowering/src/lib.rs
@@ -1285,7 +1285,7 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
     fn lower_ty_direct(&mut self, t: &Ty, itctx: ImplTraitContext) -> hir::Ty<'hir> {
         let kind = match &t.kind {
             TyKind::Infer => hir::TyKind::Infer,
-            TyKind::Err => hir::TyKind::Err(self.dcx().has_errors().unwrap()),
+            TyKind::Err(guar) => hir::TyKind::Err(*guar),
             // Lower the anonymous structs or unions in a nested lowering context.
             //
             // ```
@@ -1503,6 +1503,7 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                 );
                 hir::TyKind::Err(guar)
             }
+            TyKind::Dummy => panic!("`TyKind::Dummy` should never be lowered"),
         };
 
         hir::Ty { kind, span: self.lower_span(t.span), hir_id: self.lower_node_id(t.id) }

--- a/compiler/rustc_ast_passes/src/ast_validation.rs
+++ b/compiler/rustc_ast_passes/src/ast_validation.rs
@@ -881,7 +881,8 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                         &item.vis,
                         errors::VisibilityNotPermittedNote::TraitImpl,
                     );
-                    if let TyKind::Err = self_ty.kind {
+                    // njn: use Dummy here
+                    if let TyKind::Err(_) = self_ty.kind {
                         this.dcx().emit_err(errors::ObsoleteAuto { span: item.span });
                     }
                     if let (&Unsafe::Yes(span), &ImplPolarity::Negative(sp)) = (unsafety, polarity)

--- a/compiler/rustc_ast_pretty/src/pprust/state.rs
+++ b/compiler/rustc_ast_pretty/src/pprust/state.rs
@@ -1048,9 +1048,14 @@ impl<'a> State<'a> {
             ast::TyKind::Infer => {
                 self.word("_");
             }
-            ast::TyKind::Err => {
+            ast::TyKind::Err(_) => {
                 self.popen();
                 self.word("/*ERROR*/");
+                self.pclose();
+            }
+            ast::TyKind::Dummy => {
+                self.popen();
+                self.word("/*DUMMY*/");
                 self.pclose();
             }
             ast::TyKind::ImplicitSelf => {

--- a/compiler/rustc_expand/src/base.rs
+++ b/compiler/rustc_expand/src/base.rs
@@ -567,10 +567,13 @@ impl DummyResult {
     }
 
     /// A plain dummy type.
-    pub fn raw_ty(sp: Span, is_error: bool) -> P<ast::Ty> {
+    pub fn raw_ty(sp: Span) -> P<ast::Ty> {
+        // FIXME(nnethercote): you might expect `ast::TyKind::Dummy` to be used here, but some
+        // values produced here end up being lowered to HIR, which `ast::TyKind::Dummy` does not
+        // support, so we use an empty tuple instead.
         P(ast::Ty {
             id: ast::DUMMY_NODE_ID,
-            kind: if is_error { ast::TyKind::Err } else { ast::TyKind::Tup(ThinVec::new()) },
+            kind: ast::TyKind::Tup(ThinVec::new()),
             span: sp,
             tokens: None,
         })
@@ -611,7 +614,7 @@ impl MacResult for DummyResult {
     }
 
     fn make_ty(self: Box<DummyResult>) -> Option<P<ast::Ty>> {
-        Some(DummyResult::raw_ty(self.span, self.is_error))
+        Some(DummyResult::raw_ty(self.span))
     }
 
     fn make_arms(self: Box<DummyResult>) -> Option<SmallVec<[ast::Arm; 1]>> {

--- a/compiler/rustc_parse/src/parser/diagnostics.rs
+++ b/compiler/rustc_parse/src/parser/diagnostics.rs
@@ -46,14 +46,14 @@ use std::ops::{Deref, DerefMut};
 use thin_vec::{thin_vec, ThinVec};
 
 /// Creates a placeholder argument.
-pub(super) fn dummy_arg(ident: Ident) -> Param {
+pub(super) fn dummy_arg(ident: Ident, guar: ErrorGuaranteed) -> Param {
     let pat = P(Pat {
         id: ast::DUMMY_NODE_ID,
         kind: PatKind::Ident(BindingAnnotation::NONE, ident, None),
         span: ident.span,
         tokens: None,
     });
-    let ty = Ty { kind: TyKind::Err, span: ident.span, id: ast::DUMMY_NODE_ID, tokens: None };
+    let ty = Ty { kind: TyKind::Err(guar), span: ident.span, id: ast::DUMMY_NODE_ID, tokens: None };
     Param {
         attrs: AttrVec::default(),
         id: ast::DUMMY_NODE_ID,
@@ -1540,14 +1540,14 @@ impl<'a> Parser<'a> {
     pub(super) fn maybe_recover_from_question_mark(&mut self, ty: P<Ty>) -> P<Ty> {
         if self.token == token::Question {
             self.bump();
-            self.dcx().emit_err(QuestionMarkInType {
+            let guar = self.dcx().emit_err(QuestionMarkInType {
                 span: self.prev_token.span,
                 sugg: QuestionMarkInTypeSugg {
                     left: ty.span.shrink_to_lo(),
                     right: self.prev_token.span,
                 },
             });
-            self.mk_ty(ty.span.to(self.prev_token.span), TyKind::Err)
+            self.mk_ty(ty.span.to(self.prev_token.span), TyKind::Err(guar))
         } else {
             ty
         }
@@ -2304,8 +2304,8 @@ impl<'a> Parser<'a> {
 
     pub(super) fn recover_bad_self_param(&mut self, mut param: Param) -> PResult<'a, Param> {
         let span = param.pat.span;
-        param.ty.kind = TyKind::Err;
-        self.dcx().emit_err(SelfParamNotFirst { span });
+        let guar = self.dcx().emit_err(SelfParamNotFirst { span });
+        param.ty.kind = TyKind::Err(guar);
         Ok(param)
     }
 
@@ -2437,7 +2437,7 @@ impl<'a> Parser<'a> {
     pub(super) fn deduplicate_recovered_params_names(&self, fn_inputs: &mut ThinVec<Param>) {
         let mut seen_inputs = FxHashSet::default();
         for input in fn_inputs.iter_mut() {
-            let opt_ident = if let (PatKind::Ident(_, ident, _), TyKind::Err) =
+            let opt_ident = if let (PatKind::Ident(_, ident, _), TyKind::Err(_)) =
                 (&input.pat.kind, &input.ty.kind)
             {
                 Some(*ident)
@@ -2644,8 +2644,10 @@ impl<'a> Parser<'a> {
                         "::",
                         Applicability::MaybeIncorrect,
                     );
-                    err.emit();
-                    return Ok(GenericArg::Type(self.mk_ty(start.to(expr.span), TyKind::Err)));
+                    let guar = err.emit();
+                    return Ok(GenericArg::Type(
+                        self.mk_ty(start.to(expr.span), TyKind::Err(guar)),
+                    ));
                 } else if token::Comma == self.token.kind || self.token.kind.should_end_const_arg()
                 {
                     // Avoid the following output by checking that we consumed a full const arg:

--- a/compiler/rustc_parse/src/parser/item.rs
+++ b/compiler/rustc_parse/src/parser/item.rs
@@ -591,7 +591,23 @@ impl<'a> Parser<'a> {
         let ty_second = if self.token == token::DotDot {
             // We need to report this error after `cfg` expansion for compatibility reasons
             self.bump(); // `..`, do not add it to expected tokens
-            Some(self.mk_ty(self.prev_token.span, TyKind::Err))
+
+            // FIXME(nnethercote): AST validation later detects this
+            // `TyKind::Err` and emits an errors. So why the unchecked
+            // ErrorGuaranteed?
+            // - A `span_delayed_bug` doesn't work here, because rustfmt can
+            //   hit this path but then not hit the follow-up path in the AST
+            //   validator that issues the error, which results in ICEs.
+            // - `TyKind::Dummy` doesn't work, because it ends up reaching HIR
+            //   lowering, which results in ICEs. Changing `TyKind::Dummy` to
+            //   `TyKind::Err` during AST validation might fix that, but that's
+            //   not possible because AST validation doesn't allow mutability.
+            //
+            // #121072 will hopefully remove all this special handling of the
+            // obsolete `impl Trait for ..` and then this can go away.
+            #[allow(deprecated)]
+            let guar = rustc_errors::ErrorGuaranteed::unchecked_error_guaranteed();
+            Some(self.mk_ty(self.prev_token.span, TyKind::Err(guar)))
         } else if has_for || self.token.can_begin_type() {
             Some(self.parse_ty()?)
         } else {
@@ -2628,13 +2644,13 @@ impl<'a> Parser<'a> {
             p.recover_diff_marker();
             let snapshot = p.create_snapshot_for_diagnostic();
             let param = p.parse_param_general(req_name, first_param).or_else(|e| {
-                e.emit();
+                let guar = e.emit();
                 let lo = p.prev_token.span;
                 p.restore_snapshot(snapshot);
                 // Skip every token until next possible arg or end.
                 p.eat_to_tokens(&[&token::Comma, &token::CloseDelim(Delimiter::Parenthesis)]);
                 // Create a placeholder argument for proper arg count (issue #34264).
-                Ok(dummy_arg(Ident::new(kw::Empty, lo.to(p.prev_token.span))))
+                Ok(dummy_arg(Ident::new(kw::Empty, lo.to(p.prev_token.span)), guar))
             });
             // ...now that we've parsed the first argument, `self` is no longer allowed.
             first_param = false;
@@ -2671,8 +2687,8 @@ impl<'a> Parser<'a> {
                     return if let Some(ident) =
                         this.parameter_without_type(&mut err, pat, is_name_required, first_param)
                     {
-                        err.emit();
-                        Ok((dummy_arg(ident), TrailingToken::None))
+                        let guar = err.emit();
+                        Ok((dummy_arg(ident, guar), TrailingToken::None))
                     } else {
                         Err(err)
                     };

--- a/compiler/rustc_parse/src/parser/path.rs
+++ b/compiler/rustc_parse/src/parser/path.rs
@@ -678,8 +678,9 @@ impl<'a> Parser<'a> {
                 c.into()
             }
             Some(GenericArg::Lifetime(lt)) => {
-                self.dcx().emit_err(errors::AssocLifetime { span, lifetime: lt.ident.span });
-                self.mk_ty(span, ast::TyKind::Err).into()
+                let guar =
+                    self.dcx().emit_err(errors::AssocLifetime { span, lifetime: lt.ident.span });
+                self.mk_ty(span, ast::TyKind::Err(guar)).into()
             }
             None => {
                 let after_eq = eq.shrink_to_hi();
@@ -779,7 +780,7 @@ impl<'a> Parser<'a> {
                     // type to determine if error recovery has occurred and if the input is not a
                     // syntactically valid type after all.
                     if let ast::TyKind::Slice(inner_ty) | ast::TyKind::Array(inner_ty, _) = &ty.kind
-                        && let ast::TyKind::Err = inner_ty.kind
+                        && let ast::TyKind::Err(_) = inner_ty.kind
                         && let Some(snapshot) = snapshot
                         && let Some(expr) =
                             self.recover_unbraced_const_arg_that_can_begin_ty(snapshot)

--- a/compiler/rustc_parse/src/parser/ty.rs
+++ b/compiler/rustc_parse/src/parser/ty.rs
@@ -346,8 +346,10 @@ impl<'a> Parser<'a> {
                 AllowCVariadic::No => {
                     // FIXME(Centril): Should we just allow `...` syntactically
                     // anywhere in a type and use semantic restrictions instead?
-                    self.dcx().emit_err(NestedCVariadicType { span: lo.to(self.prev_token.span) });
-                    TyKind::Err
+                    let guar = self
+                        .dcx()
+                        .emit_err(NestedCVariadicType { span: lo.to(self.prev_token.span) });
+                    TyKind::Err(guar)
                 }
             }
         } else {
@@ -493,8 +495,8 @@ impl<'a> Parser<'a> {
             {
                 // Recover from `[LIT; EXPR]` and `[LIT]`
                 self.bump();
-                err.emit();
-                self.mk_ty(self.prev_token.span, TyKind::Err)
+                let guar = err.emit();
+                self.mk_ty(self.prev_token.span, TyKind::Err(guar))
             }
             Err(err) => return Err(err),
         };

--- a/compiler/rustc_passes/src/hir_stats.rs
+++ b/compiler/rustc_passes/src/hir_stats.rs
@@ -616,8 +616,9 @@ impl<'v> ast_visit::Visitor<'v> for StatCollector<'v> {
                 Infer,
                 ImplicitSelf,
                 MacCall,
-                Err,
-                CVarArgs
+                CVarArgs,
+                Dummy,
+                Err
             ]
         );
 

--- a/src/tools/clippy/clippy_utils/src/ast_utils.rs
+++ b/src/tools/clippy/clippy_utils/src/ast_utils.rs
@@ -690,7 +690,7 @@ pub fn eq_ty(l: &Ty, r: &Ty) -> bool {
     match (&l.kind, &r.kind) {
         (Paren(l), _) => eq_ty(l, r),
         (_, Paren(r)) => eq_ty(l, r),
-        (Never, Never) | (Infer, Infer) | (ImplicitSelf, ImplicitSelf) | (Err, Err) | (CVarArgs, CVarArgs) => true,
+        (Never, Never) | (Infer, Infer) | (ImplicitSelf, ImplicitSelf) | (Err(_), Err(_)) | (CVarArgs, CVarArgs) => true,
         (Slice(l), Slice(r)) => eq_ty(l, r),
         (Array(le, ls), Array(re, rs)) => eq_ty(le, re) && eq_expr(&ls.value, &rs.value),
         (Ptr(l), Ptr(r)) => l.mutbl == r.mutbl && eq_ty(&l.ty, &r.ty),

--- a/src/tools/rustfmt/src/types.rs
+++ b/src/tools/rustfmt/src/types.rs
@@ -859,7 +859,7 @@ impl Rewrite for ast::Ty {
                 })
             }
             ast::TyKind::CVarArgs => Some("...".to_owned()),
-            ast::TyKind::Err => Some(context.snippet(self.span).to_owned()),
+            ast::TyKind::Dummy | ast::TyKind::Err(_) => Some(context.snippet(self.span).to_owned()),
             ast::TyKind::Typeof(ref anon_const) => rewrite_call(
                 context,
                 "typeof",

